kratos/geometries/line_2d_2.h

```diff
@@ -1066,28 +1066,16 @@ class Line2D2 : public Geometry<TPointType>
         const TPointType& r_first_point  = BaseType::GetPoint(0);
         const TPointType& r_second_point = BaseType::GetPoint(1);
 
-        // Project point
-        const double tolerance = 1e-14; // Tolerance
+        // Project the point on the line in global space
+        const array_1d<double, 3> vector_from_first_point_to_input = rPoint - r_first_point;
+        const array_1d<double, 3> unity_line_direction             = (r_second_point - r_first_point) / Length();
+        const auto projection_on_line                              = inner_prod(vector_from_first_point_to_input, unity_line_direction);
 
-        const double length = Length();
+        // Conversion to local space
+        constexpr double tolerance = 1e-14;
+        rResult[0] = 2.0 * projection_on_line/(Length() + tolerance) - 1.0;
```

Comment:
No need to compute `Length()` twice.

Besides, the job of that tolerance is probably to prevent division by zero errors but there's a naked division on line 1071. I'd either get rid of `tolerance` completely (preferred - degenerate lines should be caught at construction) or apply it there as well to remain consistent.

Reply:
You are right about avoiding computing Length twice, but the tolerance is not because is degenerated, it is because the tolerance on the checks to avoid false negatives.

Reply:
False negatives?

As I understand this function doesn't do point membership tests but transforms a coordinate from global to local space.

Reply:
Yes, but that is used in the IsInside function, maybe the proper way should to check tolerance there (where Toelrance is an input)

Reply:
Yup, I like that idea better.

Reply:
By the way `IsInside` should probably be changed after this PR because an infinitely large cylinder would project to this line.

Reply:
I agree, at first I removed the tolerance completely, but it failed some test cases in different applications, so it would require quite some effort to fix (simply adding the tolerance when calling IsInside didn't work unfortunately).

I can have a look again after I fixed the issue where using auto for an array1d gave issues, maybe the tolerance can now be removed safely.

However, if that's not the case and we need to change quite a bit of code, I would suggest to do that in a separate PR and keep this one focused on just fixing the projection and keeping the tolerance that was already there.